staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise_test.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package promise
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+func TestWriteOnce(t *testing.T) {
+	oldTime := time.Now()
+	cval := &oldTime
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	wr := NewWriteOnce(nil, ctx.Done(), cval)
+	var gots int32
+	var got atomic.Value
+	go func() {
+		got.Store(wr.Get())
+		atomic.AddInt32(&gots, 1)
+	}()
+	time.Sleep(5 * time.Second)
```

Comment:
Can we avoid Sleeps?

What is the reasoning behind this? Waiting for goroutine to finish?
1) If so - then what's the point behind having a goroutine even?
2) If there is a reason that I don't see - then let's use sync.WaitGroup, add "defer wg.Done()" in the gorotuine and use wg.Wait() here.

Reply:
Same in other places.

Reply:
In this particular case, as the following check that `gots` is zero shows, the purpose of the forked goroutine and sleep is to check that the `Get` does _not_ return promptly.

In some of the other cases, the purpose of the fork & sleep is to enable the test to complete in a modest amount of time even if the forked goroutine hangs.

Reply:
> In some of the other cases, the purpose of the fork & sleep is to enable the test to complete in a modest amount of time even if the forked goroutine hangs.

In such case, the hanging is a clear bug - the pattern we use in most of places is just wait for wait.ForeverTestTimeout (i.e. the usual pattern is:
select {
case /* expected case */:
case <-time.After(wait.ForeverTestTimeout):
  t.Errorf/Fatalf(...)
}

I would really like to avoid a number Sleep(5) as this is making the test unnecessarily long (and shortening them too much can cause flakiness).